File: legacy/pasarguardbot/frontend/src/types/panel/reports.ts
/** Mirrors app/models/panel/reports.py */
import type { PagedRequest, PageMeta, PanelAuthRequest, PanelEnvelope } from "./common";

export const REPORT_PERIODS = ["today", "3d", "week", "month", "quarter", "year", "all"] as const;

export interface PanelRankRow {
  rank: number;
  user_id?: number | null;
  amount: number;
  count: number;
}

export interface PanelReportTotals {
  new_users: number;
  services_sold: number;
  test_services: number;
  manual_approved_sum: number;
  auto_approved_sum: number;
  pending_sum: number;
  pending_count: number;
}

export interface PanelReportsRequest extends PanelAuthRequest {
  /** today | 3d | week | month | quarter | year | all */
  period?: string;
}

export interface PanelReportsResponse extends PanelEnvelope {
  period: string;
  period_start: number;
  periods: string[];
  totals: PanelReportTotals;
  top_recharge: PanelRankRow[];
  top_spenders: PanelRankRow[];
  top_service_counts: PanelRankRow[];
}

export interface PanelAuditRow {
  id: number;
  actor_id?: number | null;
  actor_username?: string | null;
  action: string;
  target_type?: string | null;
  target_id?: string | null;
  detail?: Record<string, unknown> | null;
  ip?: string | null;
  created_at?: number | null;
}

export interface PanelAuditRequest extends PagedRequest {
  action?: string;
  actor_id?: number | null;
}

export interface PanelAuditResponse extends PanelEnvelope {
  entries: PanelAuditRow[];
  actions: string[];
  meta: PageMeta;
}

File: legacy/pasarguardbot/app/models/panel/reports.py
"""Admin panel DTOs: sales reports and the panel activity log."""

from typing import Any

from pydantic import BaseModel, Field

from app.models.panel.common import PagedRequest, PageMeta, PanelRequest, PanelResponse

REPORT_PERIODS = ("today", "3d", "week", "month", "quarter", "year", "all")


class PanelRankRow(BaseModel):
    rank: int
    user_id: int | None = None
    amount: int = 0
    count: int = 0


class PanelReportTotals(BaseModel):
    new_users: int = 0
    services_sold: int = 0
    test_services: int = 0
    manual_approved_sum: int = 0
    auto_approved_sum: int = 0
    pending_sum: int = 0
    pending_count: int = 0


class PanelReportsRequest(PanelRequest):
    period: str = Field("today", description="today | 3d | week | month | quarter | year | all")


class PanelReportsResponse(PanelResponse):
    period: str = "today"
    period_start: int = 0
    periods: list[str] = Field(default_factory=lambda: list(REPORT_PERIODS))
    totals: PanelReportTotals = Field(default_factory=PanelReportTotals)
    top_recharge: list[PanelRankRow] = Field(default_factory=list)
    top_spenders: list[PanelRankRow] = Field(default_factory=list)
    top_service_counts: list[PanelRankRow] = Field(default_factory=list)


class PanelAuditRow(BaseModel):
    id: int
    actor_id: int | None = None
    actor_username: str | None = None
    action: str
    target_type: str | None = None
    target_id: str | None = None
    detail: dict[str, Any] | None = None
    ip: str | None = None
    created_at: int | None = None


class PanelAuditRequest(PagedRequest):
    action: str = Field("", max_length=64)
    actor_id: int | None = None


class PanelAuditResponse(PanelResponse):
    entries: list[PanelAuditRow] = Field(default_factory=list)
    actions: list[str] = Field(default_factory=list)
    meta: PageMeta = Field(default_factory=PageMeta)
